### ISSUE TITLE
RemoteServiceException when using foreground service

### ISSUE THREAD

**Issue by mars-lan** [notifee/react-native-notifee#290]
Start seeing the following crash when rolling out our new version that uses [foreground service](https://notifee.app/react-native/docs/android/foreground-service)

```
Fatal Exception: android.app.RemoteServiceException
Context.startForegroundService() did not then call Service.startForeground()

android.app.ActivityThread$H.handleMessage (ActivityThread.java:1881)
android.os.Handler.dispatchMessage (Handler.java:105)
android.os.Looper.loop (Looper.java:164)
android.app.ActivityThread.main (ActivityThread.java:6938)
java.lang.reflect.Method.invoke (Method.java)
com.android.internal.os.Zygote$MethodAndArgsCaller.run (Zygote.java:327)
com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1374)
```

So far the crash is limited to only Android 8.0 devices but not sure if it's an artifact of the small rollout percentage or an actual version-specific issue.

Based on https://developer.android.com/about/versions/oreo/android-8.0-changes & https://developer.android.com/reference/android/app/Service#startForeground(int,%20android.app.Notification) it seems to suggest that  `startForeground` wasn't called within the 5 seconds of foreground service creation. Without any visibility into the code base, it's difficult to debug the issue further.

**Comment by mars-lan** [comment]
Can confirm that this is an Android 8 (API 26) specific issue. Was able to reproduce the crash on the simulator by calling `notifee.displayNotification` with `android.asForegroundService` set to `true`. The same code works on API 24-25, 27-30 (foreground notification somehow [isn't shown in API 22-25](https://github.com/notifee/react-native-notifee/issues/291) but at least it didn't crash)

**Comment by mikehardy** [comment]
Strange - I wonder if this is at all related to something happening with disk access (thus causing it to take longer than 5 seconds) as a proximate cause, using StrictMode to clean up things in general may be a useful pass in the codebase. I'm trying to understand (in my mental model) how a 5s violation could be API specific though, it doesn't really fit my model as I understand things. Odd.

**Comment by mars-lan** [comment]
Yeah I'm not so sure about the 5s thingy either after narrowing it down to only API 26. Since it's so readily reproducible hopefully we'll hear from the notifee team soon.

**Comment by helenaford** [comment]
@mars-lan thanks for the info and sorry you've experienced this. I had a quick look, nothing stands out what could be causing the issue. I'll run it locally with the emulator and get back to you with more details. At the very least, I can put something in place to stop the crashing (i.e. disable foreground service for api 26) like you said. Hopefully, we can resolve this though, and get the other versions working too.

**Comment by helenaford** [comment]
@mars-lan I'm struggling to recreate this issue with API 26, with the following emulator:
<img width="1477" alt="Screenshot 2021-04-27 at 17 08 40" src="https://user-images.githubusercontent.com/14185925/116275123-395acc80-a77b-11eb-8ff9-615271b89eeb.png">

Also tried with API 23:
<img width="526" alt="Screenshot 2021-04-27 at 17 14 37" src="https://user-images.githubusercontent.com/14185925/116277922-dc144a80-a77d-11eb-99e6-a6daa6ddc9e5.png">

with notification:
```
   await notifee.displayNotification({
      title: "Notification Title",
      body: "Main body content of the notification",
      android: {
        channelId,
        asForegroundService: true,
      },
    });
```

Could you can paste some example code or show what payload you are passing to `displayNotification` please?  Or, list out steps you are doing? Are you calling `displayNotification` in the foreground? Did you reproduce this in debug mode?

**Comment by mikehardy** [comment]
I wonder if it is timing dependent (the 5 second limit), and if the notification does some super slow network fetch or something it would trigger it? Is that possible - for the notification in *any* way to trigger some slow process that might take more than 5 seconds? Or maybe just for testing hack right into the code a delay to see if it reproduces the stack

**Comment by mikehardy** [comment]
Main reason I am paying a lot of attention to this is that StrictMode and 5-seconds-to-start-foreground etc - if not pursued to a conclusive end that usually results in starting a worker thread and hiving slow things off into it - will result in flaky behavior forever, and future reports that are hard to reproduce / make working with the library touchy. I've seen that happen in the other lib (react-native-background-geolocation-android) I use that does foreground service and could ANR on startup until it was thoroughly scrubbed

**Comment by helenaford** [comment]
yeah that's what i was thinking it could be, like it's downloading an image and takes a while to start the foreground service 🤔 And, maybe it happens to be more common on older API versions. So the bug doesn't necessarily relate to the api version.

**Comment by helenaford** [comment]
@mikehardy just had a look at the library you mentioned, same issue was reported here https://github.com/transistorsoft/react-native-background-geolocation/issues/976. I'm not sure though if there's something we can do in notifee to help.

**Comment by helenaford** [comment]
and here https://github.com/transistorsoft/react-native-background-geolocation/issues/810#issuecomment-810147140. very similar logs @mars-lan. is this something you could try enabling, StrictMode to help debug? it could be something blocking the main thread?

**Comment by mikehardy** [comment]
My guess it will end up being an "all of the above" approach - there are likely StrictMode violations in notifee and in @mars-lan app, and they all add up to a 5s failure

I'm saying that because I'd bet 1% of apps or less are StrictMode clean, it's a safe bet :)

**Comment by mars-lan** [comment]
@helenaford Whoa! Found the root cause! If I set `largeIcon` to a slightly large local file (256x256 25KB PNG), it'll lead to a crash. Resizing the image down to 128x128 fixed the problem. I suspect Android 8 has a slow/buggy PNG decoding that leads to the eventual 5s timeout. Here's the canonical code to reproduce it.

```
import React from 'react';
import { AppRegistry } from 'react-native';
import notifee from '@notifee/react-native';

notifee.registerForegroundService(async () => {
  return new Promise(() => {});
});

notifee.createChannel({
  id: 'channel',
  name: 'channel',
});

function Test() {
  React.useEffect(() => {
    notifee.displayNotification({
      title: 'Title',
      body: 'Body',
      android: {
        channelId: 'channel',
        largeIcon: require('./image.png'),
        asForegroundService: true,
      },
    });
  }, []);
  return null;
}

AppRegistry.registerComponent('Test', () => Test);
```

**Comment by helenaford** [comment]
ah nice. OK, so would putting this in the docs to say be cautious about image sizes be a good fix for this?

**Comment by mars-lan** [comment]
I think so. Maybe note that this seems to be an Android 8-specific issue. Another option is to move the image to Android drawable resource, which doesn't seem to suffer the same performance issue as local files.

**Comment by helenaford** [comment]
sounds good. thanks for bringing this issue to our attention and finding the bug. Would this be the same reason for the other api versions?

**Comment by mikehardy** [comment]
Maybe put the file decode into bitmap (I think Notifee is handling the file fetch and decode?) into a background thread since it is disk access and update the notification with the bitmap once it's decoded? Is that possible? All the StrictMode violations and timeout items end up with solutions like that

**Comment by mars-lan** [comment]
@helenaford I can confirm that by using an image with smaller dimension (128x128) or Android drawables directly for `largeIcon`, I was able to get the notification shown in all API versions (22-28). I did discover another somewhat-related strange bug with notifee 1.5.0 though. Will create a separate issue after more digging.

**Comment by mars-lan** [comment]
@helenaford after rolling out an update that uses Android drawable for `largeIcon`, we still observed a handful of crashes on Android 8 in production. Looks like we may need to implement a proper fix as per @mikehardy's suggestion to avoid the 5s timeout.

**Comment by mikehardy** [comment]
I need a reaction for that where it's clear I'm unhappy but not at all at you personally. :weary: - thanks for the report

**Comment by mars-lan** [comment]
FWIW also started seeing a few crashs on Android 10 (API 29). Definitely less common than API 26 though.

**Comment by mars-lan** [comment]
@helenaford I assume notifee does the following when `displayNotification` is called

1. Call [startForegroundService](https://developer.android.com/reference/android/content/Context#startForegroundService(android.content.Intent))
2. Build [Notification](https://developer.android.com/reference/android/app/Notification)
3. Call [startForeground](https://developer.android.com/reference/android/app/Service#startForeground(int,%20android.app.Notification)) with the notification

How about simply swapping 1 & 2 to minimze the time between `startForegroundService` & `startForeground`?

**Comment by helenaford** [comment]
Hey @mars-lan , I'm looking into if we can fix the image loading, not sure how it's blocking as it's done with tasks but on my TODO is to check it in 'strict' mode. 

Unfortunately, it already does that, builds the notification, then does `startForegroundService`, and `startForeground` in that order. It was one of the first things I checked 😢 I think it's about the main thread being blocked somehow with the image decoding, even after the notification is built. We pass the built notification to startForegroundService.

**Comment by mikehardy** [comment]
Sounds like it is actually doing 2 then 1 and 3 per the comment above descriptors, vs 1 3 then 2. Is it even possible to startForegroundService before you have the notification built? sounds like maybe not

**Comment by mars-lan** [comment]
@mikehardy 3 takes the notification as one of the input so 1-3-2 isn't really an option here.  

Thanks for the explanation, @helenaford. Hopefully strict mode will reveal more. At least it's consistently reproducible on Andorid 8 emulator so hopefully it makes debugging easier.

A somewhat hacky solution is to use [startForegound](https://developer.android.com/reference/android/app/Service#startForeground(int,%20android.app.Notification)) instead if `displayNotification` is called in the foreground, which I assume is the majority of the cases. This way it won't be subjected to the 5s ANR.

**Comment by mars-lan** [comment]
@helenaford  Based on https://stackoverflow.com/questions/44425584/context-startforegroundservice-did-not-then-call-service-startforeground & https://issuetracker.google.com/issues/76112072, this seemss like a common issue for Android. In fact, there's simply no way to guarantee that the system will even start the foreground service within 5s, regardless how little work there is beteween `startForegroundService` & `startForeground`. This is consistent with what we observed in the wild even after dropping `largeIcon` altogether.

Looks like the only foolproof way to avoid ANR is context.bindService + startService + service.startForeground. As a developer, I'd rather lose the ability to invoke `displayNotification` in background, then dealing with potential crashes that I can't control.

**Comment by mikehardy** [comment]
Wow! Great digging @mars-lan - lot of electrons spent discussing this one.
Looks like https://stackoverflow.com/a/57521350/9910298 is a great example of how to use bindService.

I'm a little confused on your last paragraph though - not to read too much in to it, but could your last sentence ("As a developer...") be considered a new paragraph?

To be precise, are you saying that unless this API is fixed in general ("the ability to invoke `displayNotification` in background") you'd rather not touch it, but you think the bindService path could fix it? Or, are you saying that by your read on the links you posted it is unfixable?

I ask because it seemed as though the bindService style (or even JobService maybe - but bindService easier...) would fix it.

Then you'd have displayNotification in the background and no crashes, which would be ideal. But I might be missing something you've already realized

**Comment by mars-lan** [comment]
@mikehardy Thanks for the info. The trick in https://stackoverflow.com/a/57521350/9910298 seems promsing. However,  https://stackoverflow.com/a/53286232/15875597 and other users' comments claimed that the approach doesn't work 100%. 

https://issuetracker.google.com/issues/76112072#comment158 says that the only reliable fix is to replace `startForegroundService` with `startService`, which means that you won't be able to invoke `displayNotifiation` when the app is runnning in background (aka [Headless](https://reactnative.dev/docs/headless-js-android) mode). What I meant is that I can certainly live with this limitation if that's the price to pay for stability.

**Comment by mikehardy** [comment]
Likely yes - crash issues need priority. You can see from the timing on previous comment (11 days ago) this was immediately prior to google IO and I think you've actually participated in react-native-firebase repo in the intervening time on related issues? Hopefully not mis-remembering, at any rate that pulled my full attention for a while and I'm still unwinding the stack from there so I haven't been able to turn attention back here. Will tag this up at least

**Comment by wilau2** [comment]
I think we have occurrences on the crash for android 11 as well. We are still at the beginning of our investigation, we've halted our rollout until we have a better idea of the impact and reproduction. 

![image](https://user-images.githubusercontent.com/5473183/120226072-77cd2680-c214-11eb-91ca-95a583d593bb.png)

**Comment by youssef-makhoul** [comment]
I know it is a really tricky bug to fix, but any idea when the fix can be delivered? and is there any way that we can contribute to the native side of the project?

**Comment by sambegin** [comment]
@mikehardy  sorry for poking again. We have some bandwidth on our end and could help investigating/resolving this issue. Any way we can contribute to the project? We still have occurrences of that in production and we can't really do anything about it. Not having news from notifee's side is also scary as we need to know if we'll have to implement our own solution or not. 

Could you please at least provide us with some insight on the progress of this issue ? 🙏  

Thanks a lot

**Comment by mikehardy** [comment]
I understand the desire to fix it, and I really appreciate the offer to help. Last time I asked internally we couldn't come up with a good way to extend access for collaboration but perhaps @Salakar has new ideas since then?

In prep for development here I swept through just yesterday and got it cleaned up so changes here on my part are safe (#327  #328  #326). Steady forward progress but doesn't look exciting. Thanks for your patience

**Comment by sambegin** [comment]
@Salakar @mikehardy  It's been 14 days without any news on this. We also tried to contact you via the ["Contact Us" section](https://notifee.app/contact) of your website and still no signs of answers. I will re-iterate my question again regarding the help we can provide? 

This is becoming quite time sensitive for us and we'd like to at least have the sentiment that this is not becoming an inactive library without any support. We'd like to know if someone is looking at this or not and if we can participate in the elaboration of a solution.

Also another question I would have regarding Notifee's support: Any idea when the `Extended` License including the `High priority support` could come out? It seems something we could really beneficiate... Please let us know, at least being transparent for all the people using this library 🙂 

Thank you very much!

**Comment by mikehardy** [comment]
Only speaking to this:

>  we'd like to at least have the sentiment that this is not becoming an inactive library without any support. We'd like to know if someone is looking at this or not and if we can participate in the elaboration of a solution.

Not including the core library changes, https://github.com/notifee/react-native-notifee/compare/v1.5.0...master

6 releases since this issue was logged

Obviously it would be better if this issue was fixed in one of those releases, but the library is receiving a steady amount of attention and effort, it is definitely under active development

**Comment by mars-lan** [comment]
Any progress on this one? :)

**Comment by mikehardy** [comment]
I must admit I was dreading the ping on this one - all I can offer is that the previous one from @sambegin is one of only a handful of messages I leave marked as unread in my software mail box. I have not made specific progress on it but am still slowly working towards it on the stack of todo items I've got in my Invertase queue. Cold comfort probably, but it is on the radar.

**Comment by mars-lan** [comment]
@mikehardy thanks for the transparency. I understand that there's always competing priorities. It took me nearly a year to get https://github.com/notifee/react-native-notifee/issues/126 to close so this one is still young :P

**Comment by mikehardy** [comment]
You always log the tough ones @mars-lan ! 😆
